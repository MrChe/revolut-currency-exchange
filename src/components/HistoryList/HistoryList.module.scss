.HistoryList {
  position: relative;
  list-style-type: none;
  padding: 0;
  margin: 0;
}
.HistoryListItem {
  position: relative;
  border-bottom: 1px solid rgb(74, 71, 84);;
}
.HistoryListWrapper {
  position: relative;
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.HistoryLeftPart {
  position: relative;
}
.HistoryRightPart {
  position: relative;
}

.HistoryCurrencyName {
  position: relative;
  font-size: 16px;
  font-weight: 500;
  color: rgb(74, 71, 84);
}

.HistoryTime {
  position: relative;
  font-size: 14px;
  font-weight: 500;
  color: #474854f0;
}

.BigCurrency {
  position: relative;
  font-size: 22px;
  font-weight: bold;
  color: rgb(9, 9, 9);
}

.SmallCurrency {
  position: relative;
  font-size: 14px;
  font-weight: bold;
  color: #11398ede;
}
