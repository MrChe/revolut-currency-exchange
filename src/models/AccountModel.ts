import { action, makeAutoObservable, observable } from "mobx";
import { v4 as uuidv4 } from "uuid";
// import { persistStore } from "../utils/mobx-persist.utils";

export class AccountModel {
  public id: string;
  public currency: string;
  public balance: number;
  // public history: [];
  constructor(data: Partial<AccountModel>) {
    this.id = data.currency || uuidv4();
    this.currency = data.currency || "";
    this.balance = Number(data.balance?.toFixed(2)) || 100;
    // this.history = data.history;
    makeAutoObservable(this, {
      currency: observable,
      balance: observable,
      updateBalance: action,
    });
    // persistStore(this, ["id", "currency", "balance"], "AccountModel");
  }

  public updateBalance = (balance: number): void => {
    this.balance = Number(balance.toFixed(2));
  };
}
