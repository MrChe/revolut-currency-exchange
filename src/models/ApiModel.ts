import Axios, { AxiosInstance } from "axios";
import { IRates } from "../models/ExchangeModel";
import { filterRates } from "../utils/currencies";

const APP_ID = "e285ab8ceb004923b3e423a0d0583915";

export class ApiModel {
  private client: AxiosInstance;
  constructor() {
    this.client = Axios.create({
      baseURL: `https://openexchangerates.org/api`,
      params: {
        base: "USD",
        app_id: APP_ID,
        symbols: "USD,EUR,GBP,UAH",
      },
      headers: {
        "Content-Type": "application/json",
      },
    });
  }

  public getLatestRatesRequest = async (): Promise<IRates | undefined> => {
    try {
      const response = await this.client.get(`/latest.json`);
      if (!response.data) {
        throw new Error("Not have any data");
      }
      return response.data;
    } catch (error) {
      console.error("getLatestRatesRequest", error);
    }
  };
}
