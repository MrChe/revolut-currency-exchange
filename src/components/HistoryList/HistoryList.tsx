import React from "react";
import { observer } from "mobx-react-lite";
import { useStore } from "../../models/connect";
import { AccountHistory } from "../../models/AccountModel";

import styles from "./HistoryList.module.scss";

export const HistoryList = observer(function HistoryList(): JSX.Element {
  const { ExchangeModel } = useStore();

  const accountHistory = ExchangeModel.activeAccountFrom?.history;

  return accountHistory && accountHistory?.length !== 0 ? (
    <ul className={styles.HistoryList}>
      {accountHistory.map((h: AccountHistory) => {
        return (
          <li key={h.id} className={styles.HistoryListItem}>
            <div className={styles.HistoryWrapper}>
              <div className={styles.HistoryLeftPart}>
                <p className={styles.HistoryCurrencyName}>{h.name}</p>
                <p className={styles.HistoryTime}>{h.time}</p>
              </div>
              <div className={styles.HistoryRightPart}>
                <p className={styles.BigCurrency}>{h.from}</p>
                <p className={styles.SmallCurrency}>{h.to}</p>
              </div>
            </div>
          </li>
        );
      })}
    </ul>
  ) : (
    <div></div>
  );
});
