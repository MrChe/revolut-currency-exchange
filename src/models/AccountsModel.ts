import { RootModel } from "./RootModel";
// import { persistStore } from "../utils/mobx-persist.utils";
import { action, computed, makeAutoObservable, observable } from "mobx";
import { AccountModel } from "./AccountModel";
import { IRates } from "./ExchangeModel";

export class AccountsModel {
  rootModel: RootModel;
  public accounts: AccountModel[];
  public ratesData: IRates | null;
  public selectedAccount: AccountModel | null;
  constructor(rootModel: RootModel, currencies: string[], base: string) {
    this.rootModel = rootModel;
    this.selectedAccount = null;
    this.accounts = [];
    this.ratesData = null;
    makeAutoObservable(this, {
      // rootModel: false,
      accounts: observable,
      selectedAccount: observable,
      accountsAsArray: computed,
      setSelectedAccount: action,
    });

    // INIT Module
    this.init(currencies, base);
    // persistStore(
    //   this,
    //   ["accounts", "selectedAccount", "accountsAsArray"],
    //   "AccountsModel",
    // );
  }

  private init = async (currencies: string[], base: string): Promise<void> => {
    try {
      if (currencies.length !== 0) {
        const data = await this.rootModel.ApiModel.getLatestRatesRequest(
          currencies,
          base,
        );
        if (data) {
          this.setRatesData(data);
          this.setAccounts(data);
        }
      } else {
        throw new Error("Please add some currencies like ['USD', 'UAH']");
      }
    } catch (error) {
      console.error("init", error);
    }
  };

  public setSelectedAccount = (id: string): void => {
    const foundAccount = this.getAccountById(id);
    if (foundAccount) {
      this.selectedAccount = foundAccount;
    } else {
      console.error(`Cant find account by id: ${id}`);
    }
  };

  public get accountsAsArray(): AccountModel[] {
    return this.accounts.slice();
  }

  public setRatesData = (ratesData: IRates): void => {
    this.ratesData = ratesData;
  };

  public setAccounts = (ratesData: IRates): void => {
    Object.keys(ratesData.rates).forEach((currency: string) => {
      this.accounts.push(
        new AccountModel({
          currency,
          balance: 100,
        }),
      );
    });
  };

  public getAccountById = (id: string): AccountModel | undefined => {
    return this.accounts.find((account) => account.id === id);
  };
}
