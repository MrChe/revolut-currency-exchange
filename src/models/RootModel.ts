import { ExchangeModel } from "./ExchangeModel";
import { AccountsModel } from "./AccountsModel";
import { ApiModel } from "./ApiModel";

export class RootModel {
  public ExchangeModel: ExchangeModel;
  public AccountsModel: AccountsModel;
  public ApiModel: ApiModel;
  constructor() {
    this.ApiModel = new ApiModel();
    // init accounts
    this.AccountsModel = new AccountsModel(
      this,
      ["USD", "EUR", "GBP", "UAH"],
      "USD",
    );
    this.ExchangeModel = new ExchangeModel(this);
  }
}

export const rootModel = new RootModel();
