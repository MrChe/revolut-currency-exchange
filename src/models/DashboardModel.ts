import { RootModel } from "./RootModel";
// import { persistStore } from "../utils/mobx-persist.utils";
import { action, makeAutoObservable, observable } from "mobx";
import { AccountModel } from "./AccountModel";
import { IRates } from "./ExchangeModel";
// import { ActiveAccounts } from "./ActiveAccountsModel";

export class DashboardModel {
  private rootModel: RootModel;
  public accounts: AccountModel[];
  public ratesData: IRates | null;
  constructor(rootModel: RootModel) {
    this.rootModel = rootModel;
    makeAutoObservable(this, {
      accounts: observable,
      setActiveAccounts: action,
    });
    this.accounts = [];
    this.ratesData = null;
    // persistStore(this, ["dashboard"], "DashboardModel");
  }

  public setRatesData = (ratesData: IRates): void => {
    this.ratesData = ratesData;
  };

  public setAccounts = (ratesData: IRates): void => {
    Object.keys(ratesData.rates).forEach((currency: string) => {
      this.accounts.push(
        new AccountModel({
          currency,
          balance: 100,
        }),
      );
    });
  };

  public getAccount = (currency: string): AccountModel | undefined => {
    return this.accounts.find((account) => account.currency === currency);
  };

  public getLatestRates = async (): Promise<void> => {
    try {
      const data = await this.rootModel.ApiModel.getLatestRatesRequest();
      if (data) {
        this.setRatesData(data);
        // this.initCashify(data);
        this.setAccounts(data);
        // this.setActiveAccounts({
        //   from: this.findAccountByCurrency("USD"),
        //   to: this.findAccountByCurrency("EUR"),
        // });
      }
    } catch (error) {
      console.error("getLatestRates", error);
    }
  };
}
