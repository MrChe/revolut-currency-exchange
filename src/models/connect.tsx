// holds a reference to the store (singleton)
import { createContext, useContext } from "react";
import { RootModel } from "./RootModel";

// create the context
const MobxContext = createContext<RootModel | null>(null);

// create the provider component
export const MobxProvider = MobxContext.Provider;

// create the hook
export const useStore = (): RootModel => {
  const context = useContext(MobxContext);
  if (!context) {
    throw new Error("useStore must be used within MobxContext");
  }

  return context;
};
