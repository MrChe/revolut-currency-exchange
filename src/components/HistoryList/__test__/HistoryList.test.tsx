import React from "react";
import renderer from "react-test-renderer";
import { HistoryList } from "../HistoryList";
import { rootModel } from "../../../models/RootModel";
import { ratesData } from "../../../../tests/mockData";
import { TestWrapper } from "../../../../tests/testWrapper";

describe("HistoryList Component", () => {
  const ExchangeModel = rootModel.ExchangeModel;
  beforeAll(() => {
    ExchangeModel.setRatesData(ratesData);
    ExchangeModel.setAccounts(ratesData);
    ExchangeModel.setActiveAccountFrom(ExchangeModel.accounts[0].id);
    ExchangeModel.setActiveAccountTo(ExchangeModel.accounts[1].id);
    if (ExchangeModel.activeAccountFrom) {
      ExchangeModel.updateHistoryInAccount(
        ExchangeModel.activeAccountFrom?.id,
        {
          name: `Exchange to ${ExchangeModel.activeAccountTo?.currency}`,
          from: `- ${ExchangeModel.formatCurrency(
            ExchangeModel.inputFromValue,
            ExchangeModel.activeAccountFrom?.currency || "",
          )}`,
          to: `+ ${ExchangeModel.formatCurrency(
            ExchangeModel.inputToValue,
            ExchangeModel.activeAccountTo?.currency || "",
          )}`,
        },
      );
    }
  });
  it("HistoryList snapshot", () => {
    const tree = renderer
      .create(
        <TestWrapper>
          <HistoryList />
        </TestWrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
