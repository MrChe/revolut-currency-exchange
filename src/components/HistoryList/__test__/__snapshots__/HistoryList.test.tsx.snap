// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryList Component HistoryList snapshot 1`] = `
<ul
  className="HistoryList"
>
  <li
    className="HistoryListItem"
  >
    <div
      className="HistoryListWrapper"
    >
      <div
        className="HistoryLeftPart"
      >
        <p
          className="HistoryCurrencyName"
        >
          Exchange to GBP
        </p>
        <p
          className="HistoryTime"
        >
          01/01/2021 00:00
        </p>
      </div>
      <div
        className="HistoryRightPart"
      >
        <p
          className="BigCurrency"
        >
          - €0.00
        </p>
        <p
          className="SmallCurrency"
        >
          + £0.00
        </p>
      </div>
    </div>
  </li>
</ul>
`;
